### DISCUSSION TITLE
.NET Framework v4.7.2 w/ mapperly v2.7.0 | nullable reference types

### DISCUSSION THREAD

**Discussion by tblong** [riok/mapperly#328] [Q&A]
![image](https://user-images.githubusercontent.com/12371929/231515226-ae06b1e7-04be-4f32-8975-7bc55ab31f1a.png)

The mapperly lib is advertising itself as .NET Standard v2.0 however I'm seeing the error `CS8370 - Feature 'nullable reference types' is not available in C# 7.3. Please use language version 8.0 or greater.` in a simple console app targeting .NET Framework v4.7.2.

Am I just out of luck getting mapperly to work with older projects where C# v7.3 is the norm?  Thanks for any guidance.

**Comment by latonz** [comment]
Mapperly supports the .NET Framework but requires C# language version 8. We'll improve the documentation on this.

You should be able to use C# 8 in your project. See https://stackoverflow.com/questions/56651472/does-c-sharp-8-support-the-net-framework for more information

**Comment by tblong** [comment]
@latonz That's helpful thanks and I have seen that post.  It does not however seem like a good practice to follow but I have gotten that to work by setting `langversion` to 8.0+ in the project file.

Source you undoubtedly know about then: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version

Even if the docs are updated for mapperly, I consider any library that advertises itself as .NET Standard v2.0 should be compatible with and generate code which is also within the scope of the C# v7.3 version and no higher.  There are also certain aspects within mapperly that require use of the C# v9.0 language version.

**Reply by latonz** [reply]
I'm aware that the situation is not optimal. Where did you find that Mapperly is .NET Standard 2.0 compatible? I thought we didn't have anything documented about compatibility yet (which we just improved with https://github.com/riok/mapperly/pull/334). I'd probably remove hints which state that Mapperly is .NET Standard 2.0 compatible to address your point as the current situation cannot be expressed as 100% compatible with .NET Standard 2.0. If you're referring to the NuGet page: we cannot really influence that. Source generators need to target .NET Standard 2.0 themselves, but this does not mean that the code generated by a source generator is 100% compatible with .NET Standard 2.0 respectively the language version in use.

I'm not aware of features that require C# v9, could you list the functions you have in mind?

**Reply by tblong** [reply]
> Where did you find that Mapperly is .NET Standard 2.0 compatible?
- https://github.com/riok/mapperly/blob/main/src/Directory.Build.props, the target framework there ends up getting used as the nuspec dependency target.  So this is really where the unexpected behavior comes from since the dependency ultimately advertised by the nuget package becomes standard v2.0.

> I'm not aware of features that require C# v9, could you list the functions you have in mind?

Certainly.  From [the getting started example](https://mapperly.riok.app/docs/getting-started/first-mapper), here is the same project where I've added `<LangVersion>8.0</LangVersion>` to the csproj file of the .NET Framework v4.7.2 targeting console app:
![image](https://user-images.githubusercontent.com/12371929/232047946-31b89a0d-5c0a-44a5-b074-33481eb052dd.png)

The error I think is referring to this C# v9.0 feature: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/extending-partial-methods

**Reply by latonz** [reply]
You are totally correct, Mapperly needs at least C# v9.0. I wasn't aware that C# only introduced the extending partial methods with v9.0. You should still be able to use the [void mapping methods](https://mapperly.riok.app/docs/configuration/void-mapping-methods)  with c# v8.0.
